More options for `sample_numpyro_nuts`: chain_method and jit placement

martiningram opened the pull request:
Hi all,

further to my comment in https://github.com/pymc-devs/pymc3/issues/4288, I thought I'd go ahead and make the pull request. Only a few lines are changed, essentially amounting to two new optional kwargs to `sample_numpyro_nuts`. The defaults match the current ones, so there should be no changes to default behaviour. The motivation is explained in the linked issue. One key fix is that it should allow the use of `chain_method='vectorized'` on the GPU, which can be faster than `chain_method='parallel'` (again see issue for an example) and also shows a progress bar.

Keen to hear people's thoughts!

Thanks,
Martin

martiningram committed:
Updates to numpyro sampler.

martiningram committed:
Optionally jit either _sample or log_p_fn

martiningram committed:
Change default to match current default

twiecki commented:
Thanks @martiningram!

There's still a failure:
```
=================================== FAILURES ===================================
____________________________ test_transform_samples ____________________________

    def test_transform_samples():
    
        with pm.Model() as model:
    
            sigma = pm.HalfNormal("sigma")
            b = pm.Normal("b", sigma=sigma)
>           trace = sample_numpyro_nuts(keep_untransformed=True)

pymc3/tests/test_sampling_jax.py:14: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

draws = 1000, tune = 1000, chains = 4, target_accept = 0.8, random_seed = 10
model = <pymc3.model.Model object at 0x7fd9ac479af0>, progress_bar = True
chain_method = 'parallel', keep_untransformed = True, jit_sample_fn = True

    def sample_numpyro_nuts(
        draws=1000,
        tune=1000,
        chains=4,
        target_accept=0.8,
        random_seed=10,
        model=None,
        progress_bar=True,
        chain_method="parallel",
        keep_untransformed=False,
        jit_sample_fn=True,
    ):
        from numpyro.infer import MCMC, NUTS
    
        from pymc3 import modelcontext
    
        model = modelcontext(model)
    
        seed = jax.random.PRNGKey(random_seed)
    
        fgraph = aesara.graph.fg.FunctionGraph(model.free_RVs, [model.logpt])
        fns = jax_funcify(fgraph)
    
>       logp_fn_jax = jax.jit(fns[0]) if not jit_sample_fn else fns[0]
E       TypeError: 'function' object is not subscriptable
```

Also would be good to add this to the tests.

martiningram commented:
Thanks for taking a look Thomas! I'm actually a little unsure what to do about this one. The original source code has:

```python
logp_fn_jax = fns[0]
```

which I have replaced with

```python
logp_fn_jax = jax.jit(fns[0]) if not jit_sample_fn else fns[0]
```

Both of them try to "subscript" the `fns` object, so I would expect the current version of pymc3 to have the exact same problem. Is this perhaps some new behaviour from `aesara`?

twiecki commented:
I don't think so, and then it wouldn't keep passing on the current tests. But I'm a bit stumped as well. Does it work locally for you?

martiningram commented:
Hi Thomas,

Really sorry for the long delay getting back to you, last week was very busy!

I tried this with different versions of `aesara`, with the following results:

Commit 4cda2e52cfa61fd3d924231ec9763956027aecc9 works. This is tag: rel-2.0.6.

Commit 1549649f0888a821b7f4c4f023c3b6234ce7ddcf or tag rel-2.0.7 however produces the expected error:

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<timed exec> in <module>

~/projects/pymc3/pymc3/sampling_jax.py in sample_numpyro_nuts(draws, tune, chains, target_accept, random_seed, model, progress_bar, chain_method, keep_untransformed, jit_sample_fn)
    139     fns = jax_funcify(fgraph)
    140 
--> 141     logp_fn_jax = jax.jit(fns[0]) if not jit_sample_fn else fns[0]
    142 
    143     rv_names = [rv.name for rv in model.free_RVs]

TypeError: 'function' object is not subscriptable
```

This suggests it broke between releases 2.0.6 and 2.0.7. I used `git bisect` to find that apparently:

```
c9333bcff533bd1a9f0543f5fa3e16b844f09c11 is the first bad commit
commit c9333bcff533bd1a9f0543f5fa3e16b844f09c11
Author: Brandon T. Willard <brandonwillard@users.noreply.github.com>
Date:   Sun Apr 11 19:15:13 2021 -0500

    Create flat functions via AST for JAXification of FunctionGraphs

 aesara/link/jax/jax_dispatch.py | 211 ++++++++++++++++++++--------------------
 aesara/link/jax/jax_linker.py   |  86 ++++++++--------
 tests/link/test_jax.py          |  93 +++++++++++++++++-
 3 files changed, 237 insertions(+), 153 deletions(-)
```
I hope that's helpful.

By the way, using the latest commit I have downloaded gives yet another error:

Commit 5fbaecc393e859c6c316789623609455cb573a48:

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~/projects/aesara/aesara/scalar/basic.py in filter(self, data, strict, allow_downcast)
    362                 )
--> 363                 or np.array_equal(data, converted_data, equal_nan=True)
    364             ):

<__array_function__ internals> in array_equal(*args, **kwargs)

TypeError: _array_equal_dispatcher() got an unexpected keyword argument 'equal_nan'
```

twiecki commented:
Good hunting! Can you try with `fns` instead of `fns[0]`?

twiecki commented:
Link to commit: https://github.com/pymc-devs/aesara/commit/c9333bcff533bd1a9f0543f5fa3e16b844f09c11

CC @brandonwillard

martiningram commented:
> Good hunting! Can you try with `fns` instead of `fns[0]`?

Sure! The new line is:

`logp_fn_jax = jax.jit(fns) if not jit_sample_fn else fns`

This seems to give a new error:

```
---------------------------------------------------------------------------
FilteredStackTrace                        Traceback (most recent call last)
<timed exec> in <module>

~/projects/pymc3/pymc3/sampling_jax.py in sample_numpyro_nuts(draws, tune, chains, target_accept, random_seed, model, progress_bar, chain_method, keep_untransformed, jit_sample_fn)
    178     map_seed = jax.random.split(seed, chains)
--> 179     mcmc_samples, leapfrogs_taken = _sample(init_state_batched, map_seed)
    180     # map_seed = jax.random.split(seed, chains)

~/projects/pymc3/pymc3/sampling_jax.py in _sample(current_state, seed)
    168 
--> 169         pmap_numpyro.run(seed, init_params=current_state, extra_fields=("num_steps",))
    170         samples = pmap_numpyro.get_samples(group_by_chain=True)

~/software/numpyro/numpyro/infer/mcmc.py in run(self, rng_key, extra_fields, init_params, *args, **kwargs)
    481                 assert chain_method == 'vectorized'
--> 482                 states, last_state = partial_map_fn(map_args)
    483                 # swap num_samples x num_chains to num_chains x num_samples

~/software/numpyro/numpyro/infer/mcmc.py in _single_chain_mcmc(self, init, args, kwargs, collect_fields)
    327             init_state = self.sampler.init(rng_key, self.num_warmup, init_params,
--> 328                                            model_args=args, model_kwargs=kwargs)
    329         sample_fn, postprocess_fn = self._get_cached_fns()

~/software/numpyro/numpyro/infer/hmc.py in init(self, rng_key, num_warmup, init_params, model_args, model_kwargs)
    477             # wa_steps because those variables do not depend on traced args: init_params, rng_key.
--> 478             init_state = vmap(hmc_init_fn)(init_params, rng_key)
    479             sample_fn = vmap(self._sample_fn, in_axes=(0, None, None))

~/software/numpyro/numpyro/infer/hmc.py in <lambda>(init_params, rng_key)
    469             model_kwargs=model_kwargs,
--> 470             rng_key=rng_key,
    471         )

~/software/numpyro/numpyro/infer/hmc.py in init_kernel(init_params, num_warmup, step_size, inverse_mass_matrix, adapt_step_size, adapt_mass_matrix, dense_mass, target_accept_prob, trajectory_length, max_tree_depth, find_heuristic_step_size, model_args, model_kwargs, rng_key)
    239         vv_init, vv_update = velocity_verlet(pe_fn, kinetic_fn)
--> 240         vv_state = vv_init(z, r, potential_energy=pe, z_grad=z_grad)
    241         energy = kinetic_fn(wa_state.inverse_mass_matrix, vv_state.r)

~/software/numpyro/numpyro/infer/hmc_util.py in init_fn(z, r, potential_energy, z_grad)
    193         if potential_energy is None or z_grad is None:
--> 194             potential_energy, z_grad = value_and_grad(potential_fn)(z)
    195         return IntegratorState(z, r, potential_energy, z_grad)

~/projects/pymc3/pymc3/sampling_jax.py in <lambda>(x)
    150         nuts_kernel = NUTS(
--> 151             potential_fn=lambda x: -logp_fn_jax(*x),
    152             # model=model,

FilteredStackTrace: TypeError: bad operand type for unary -: 'tuple'

The stack trace above excludes JAX-internal frames.
The following is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

TypeError                                 Traceback (most recent call last)
<timed exec> in <module>

~/projects/pymc3/pymc3/sampling_jax.py in sample_numpyro_nuts(draws, tune, chains, target_accept, random_seed, model, progress_bar, chain_method, keep_untransformed, jit_sample_fn)
    177     tic2 = pd.Timestamp.now()
    178     map_seed = jax.random.split(seed, chains)
--> 179     mcmc_samples, leapfrogs_taken = _sample(init_state_batched, map_seed)
    180     # map_seed = jax.random.split(seed, chains)
    181     # mcmc_samples = _sample(init_state_batched, map_seed)

~/projects/pymc3/pymc3/sampling_jax.py in _sample(current_state, seed)
    167         )
    168 
--> 169         pmap_numpyro.run(seed, init_params=current_state, extra_fields=("num_steps",))
    170         samples = pmap_numpyro.get_samples(group_by_chain=True)
    171         leapfrogs_taken = pmap_numpyro.get_extra_fields(group_by_chain=True)["num_steps"]

~/software/numpyro/numpyro/infer/mcmc.py in run(self, rng_key, extra_fields, init_params, *args, **kwargs)
    480             else:
    481                 assert chain_method == 'vectorized'
--> 482                 states, last_state = partial_map_fn(map_args)
    483                 # swap num_samples x num_chains to num_chains x num_samples
    484                 states = tree_map(lambda x: jnp.swapaxes(x, 0, 1), states)

~/software/numpyro/numpyro/infer/mcmc.py in _single_chain_mcmc(self, init, args, kwargs, collect_fields)
    326         if init_state is None:
    327             init_state = self.sampler.init(rng_key, self.num_warmup, init_params,
--> 328                                            model_args=args, model_kwargs=kwargs)
    329         sample_fn, postprocess_fn = self._get_cached_fns()
    330         diagnostics = lambda x: self.sampler.get_diagnostics_str(x[0]) if rng_key.ndim == 1 else ''   # noqa: E731

~/software/numpyro/numpyro/infer/hmc.py in init(self, rng_key, num_warmup, init_params, model_args, model_kwargs)
    476             # nonlocal variables: momentum_generator, wa_update, trajectory_len, max_treedepth,
    477             # wa_steps because those variables do not depend on traced args: init_params, rng_key.
--> 478             init_state = vmap(hmc_init_fn)(init_params, rng_key)
    479             sample_fn = vmap(self._sample_fn, in_axes=(0, None, None))
    480             self._sample_fn = sample_fn

~/miniconda3/envs/pymc3/lib/python3.7/site-packages/jax/_src/traceback_util.py in reraise_with_filtered_traceback(*args, **kwargs)
    137   def reraise_with_filtered_traceback(*args, **kwargs):
    138     try:
--> 139       return fun(*args, **kwargs)
    140     except Exception as e:
    141       if not is_under_reraiser(e):

~/miniconda3/envs/pymc3/lib/python3.7/site-packages/jax/api.py in batched_fun(*args, **kwargs)
   1230         flat_fun, axis_name, axis_size, in_axes_flat,
   1231         lambda: flatten_axes("vmap out_axes", out_tree(), out_axes)
-> 1232     ).call_wrapped(*args_flat)
   1233     return tree_unflatten(out_tree(), out_flat)
   1234 

~/miniconda3/envs/pymc3/lib/python3.7/site-packages/jax/linear_util.py in call_wrapped(self, *args, **kwargs)
    164 
    165     try:
--> 166       ans = self.f(*args, **dict(self.params, **kwargs))
    167     except:
    168       # Some transformations yield from inside context managers, so we have to

~/software/numpyro/numpyro/infer/hmc.py in <lambda>(init_params, rng_key)
    468             model_args=model_args,
    469             model_kwargs=model_kwargs,
--> 470             rng_key=rng_key,
    471         )
    472         if rng_key.ndim == 1:

~/software/numpyro/numpyro/infer/hmc.py in init_kernel(init_params, num_warmup, step_size, inverse_mass_matrix, adapt_step_size, adapt_mass_matrix, dense_mass, target_accept_prob, trajectory_length, max_tree_depth, find_heuristic_step_size, model_args, model_kwargs, rng_key)
    238         r = momentum_generator(z, wa_state.mass_matrix_sqrt, rng_key_momentum)
    239         vv_init, vv_update = velocity_verlet(pe_fn, kinetic_fn)
--> 240         vv_state = vv_init(z, r, potential_energy=pe, z_grad=z_grad)
    241         energy = kinetic_fn(wa_state.inverse_mass_matrix, vv_state.r)
    242         hmc_state = HMCState(0, vv_state.z, vv_state.z_grad, vv_state.potential_energy, energy,

~/software/numpyro/numpyro/infer/hmc_util.py in init_fn(z, r, potential_energy, z_grad)
    192         """
    193         if potential_energy is None or z_grad is None:
--> 194             potential_energy, z_grad = value_and_grad(potential_fn)(z)
    195         return IntegratorState(z, r, potential_energy, z_grad)
    196 

~/miniconda3/envs/pymc3/lib/python3.7/site-packages/jax/_src/traceback_util.py in reraise_with_filtered_traceback(*args, **kwargs)
    137   def reraise_with_filtered_traceback(*args, **kwargs):
    138     try:
--> 139       return fun(*args, **kwargs)
    140     except Exception as e:
    141       if not is_under_reraiser(e):

~/miniconda3/envs/pymc3/lib/python3.7/site-packages/jax/api.py in value_and_grad_f(*args, **kwargs)
    813     tree_map(partial(_check_input_dtype_grad, holomorphic, allow_int), dyn_args)
    814     if not has_aux:
--> 815       ans, vjp_py = _vjp(f_partial, *dyn_args)
    816     else:
    817       ans, vjp_py, aux = _vjp(f_partial, *dyn_args, has_aux=True)

~/miniconda3/envs/pymc3/lib/python3.7/site-packages/jax/api.py in _vjp(fun, has_aux, *primals)
   1886   if not has_aux:
   1887     flat_fun, out_tree = flatten_fun_nokwargs(fun, in_tree)
-> 1888     out_primal, out_vjp = ad.vjp(flat_fun, primals_flat)
   1889     out_tree = out_tree()
   1890   else:

~/miniconda3/envs/pymc3/lib/python3.7/site-packages/jax/interpreters/ad.py in vjp(traceable, primals, has_aux)
    112 def vjp(traceable, primals, has_aux=False):
    113   if not has_aux:
--> 114     out_primals, pvals, jaxpr, consts = linearize(traceable, *primals)
    115   else:
    116     out_primals, pvals, jaxpr, consts, aux = linearize(traceable, *primals, has_aux=True)

~/miniconda3/envs/pymc3/lib/python3.7/site-packages/jax/interpreters/ad.py in linearize(traceable, *primals, **kwargs)
     99   _, in_tree = tree_flatten(((primals, primals), {}))
    100   jvpfun_flat, out_tree = flatten_fun(jvpfun, in_tree)
--> 101   jaxpr, out_pvals, consts = pe.trace_to_jaxpr(jvpfun_flat, in_pvals)
    102   out_primals_pvals, out_tangents_pvals = tree_unflatten(out_tree(), out_pvals)
    103   assert all(out_primal_pval.is_known() for out_primal_pval in out_primals_pvals)

~/miniconda3/envs/pymc3/lib/python3.7/site-packages/jax/interpreters/partial_eval.py in trace_to_jaxpr(fun, pvals, instantiate, stage_out, bottom, trace_type)
   1305     with core.new_main(trace_type, bottom=bottom) as main:
   1306       fun = trace_to_subjaxpr(fun, main, instantiate)
-> 1307       jaxpr, (out_pvals, consts, env) = fun.call_wrapped(pvals)
   1308       assert not env
   1309       del main

~/miniconda3/envs/pymc3/lib/python3.7/site-packages/jax/linear_util.py in call_wrapped(self, *args, **kwargs)
    164 
    165     try:
--> 166       ans = self.f(*args, **dict(self.params, **kwargs))
    167     except:
    168       # Some transformations yield from inside context managers, so we have to

~/projects/pymc3/pymc3/sampling_jax.py in <lambda>(x)
    149         step_size = jax.tree_map(jax.numpy.ones_like, init_state)
    150         nuts_kernel = NUTS(
--> 151             potential_fn=lambda x: -logp_fn_jax(*x),
    152             # model=model,
    153             target_accept_prob=target_accept,

TypeError: bad operand type for unary -: 'tuple'
```

I was concerned that this might be because my version of numpyro is old. I reinstalled to version 0.6.0 and now get another error:

```
---------------------------------------------------------------------------
FilteredStackTrace                        Traceback (most recent call last)
<timed exec> in <module>

~/projects/pymc3/pymc3/sampling_jax.py in sample_numpyro_nuts(draws, tune, chains, target_accept, random_seed, model, progress_bar, chain_method, keep_untransformed, jit_sample_fn)
    178     map_seed = jax.random.split(seed, chains)
--> 179     mcmc_samples, leapfrogs_taken = _sample(init_state_batched, map_seed)
    180     # map_seed = jax.random.split(seed, chains)

~/projects/pymc3/pymc3/sampling_jax.py in _sample(current_state, seed)
    168 
--> 169         pmap_numpyro.run(seed, init_params=current_state, extra_fields=("num_steps",))
    170         samples = pmap_numpyro.get_samples(group_by_chain=True)

~/miniconda3/envs/pymc3/lib/python3.7/site-packages/numpyro/infer/mcmc.py in run(self, rng_key, extra_fields, init_params, *args, **kwargs)
    506                 assert self.chain_method == 'vectorized'
--> 507                 states, last_state = partial_map_fn(map_args)
    508                 # swap num_samples x num_chains to num_chains x num_samples

~/miniconda3/envs/pymc3/lib/python3.7/site-packages/numpyro/infer/mcmc.py in _single_chain_mcmc(self, init, args, kwargs, collect_fields)
    333             init_state = self.sampler.init(rng_key, self.num_warmup, init_params,
--> 334                                            model_args=args, model_kwargs=kwargs)
    335         sample_fn, postprocess_fn = self._get_cached_fns()

~/miniconda3/envs/pymc3/lib/python3.7/site-packages/numpyro/infer/hmc.py in init(self, rng_key, num_warmup, init_params, model_args, model_kwargs)
    509             # wa_steps because those variables do not depend on traced args: init_params, rng_key.
--> 510             init_state = vmap(hmc_init_fn)(init_params, rng_key)
    511             sample_fn = vmap(self._sample_fn, in_axes=(0, None, None))

~/miniconda3/envs/pymc3/lib/python3.7/site-packages/numpyro/infer/hmc.py in <lambda>(init_params, rng_key)
    501             model_kwargs=model_kwargs,
--> 502             rng_key=rng_key,
    503         )

~/miniconda3/envs/pymc3/lib/python3.7/site-packages/numpyro/infer/hmc.py in init_kernel(init_params, num_warmup, step_size, inverse_mass_matrix, adapt_step_size, adapt_mass_matrix, dense_mass, target_accept_prob, trajectory_length, max_tree_depth, find_heuristic_step_size, forward_mode_differentiation, model_args, model_kwargs, rng_key)
    210         step_size = lax.convert_element_type(step_size, jnp.result_type(float))
--> 211         trajectory_length = lax.convert_element_type(trajectory_length, jnp.result_type(float))
    212         nonlocal wa_update, max_treedepth, vv_update, wa_steps, forward_mode_ad

FilteredStackTrace: TypeError: Argument 'None' of type '<class 'NoneType'>' is not a valid JAX type

The stack trace above excludes JAX-internal frames.
The following is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

TypeError                                 Traceback (most recent call last)
<timed exec> in <module>

~/projects/pymc3/pymc3/sampling_jax.py in sample_numpyro_nuts(draws, tune, chains, target_accept, random_seed, model, progress_bar, chain_method, keep_untransformed, jit_sample_fn)
    177     tic2 = pd.Timestamp.now()
    178     map_seed = jax.random.split(seed, chains)
--> 179     mcmc_samples, leapfrogs_taken = _sample(init_state_batched, map_seed)
    180     # map_seed = jax.random.split(seed, chains)
    181     # mcmc_samples = _sample(init_state_batched, map_seed)

~/projects/pymc3/pymc3/sampling_jax.py in _sample(current_state, seed)
    167         )
    168 
--> 169         pmap_numpyro.run(seed, init_params=current_state, extra_fields=("num_steps",))
    170         samples = pmap_numpyro.get_samples(group_by_chain=True)
    171         leapfrogs_taken = pmap_numpyro.get_extra_fields(group_by_chain=True)["num_steps"]

~/miniconda3/envs/pymc3/lib/python3.7/site-packages/numpyro/infer/mcmc.py in run(self, rng_key, extra_fields, init_params, *args, **kwargs)
    505             else:
    506                 assert self.chain_method == 'vectorized'
--> 507                 states, last_state = partial_map_fn(map_args)
    508                 # swap num_samples x num_chains to num_chains x num_samples
    509                 states = tree_map(lambda x: jnp.swapaxes(x, 0, 1), states)

~/miniconda3/envs/pymc3/lib/python3.7/site-packages/numpyro/infer/mcmc.py in _single_chain_mcmc(self, init, args, kwargs, collect_fields)
    332         if init_state is None:
    333             init_state = self.sampler.init(rng_key, self.num_warmup, init_params,
--> 334                                            model_args=args, model_kwargs=kwargs)
    335         sample_fn, postprocess_fn = self._get_cached_fns()
    336         diagnostics = lambda x: self.sampler.get_diagnostics_str(x[0]) if rng_key.ndim == 1 else ''   # noqa: E731

~/miniconda3/envs/pymc3/lib/python3.7/site-packages/numpyro/infer/hmc.py in init(self, rng_key, num_warmup, init_params, model_args, model_kwargs)
    508             # nonlocal variables: momentum_generator, wa_update, trajectory_len, max_treedepth,
    509             # wa_steps because those variables do not depend on traced args: init_params, rng_key.
--> 510             init_state = vmap(hmc_init_fn)(init_params, rng_key)
    511             sample_fn = vmap(self._sample_fn, in_axes=(0, None, None))
    512             self._sample_fn = sample_fn

~/miniconda3/envs/pymc3/lib/python3.7/site-packages/jax/_src/traceback_util.py in reraise_with_filtered_traceback(*args, **kwargs)
    137   def reraise_with_filtered_traceback(*args, **kwargs):
    138     try:
--> 139       return fun(*args, **kwargs)
    140     except Exception as e:
    141       if not is_under_reraiser(e):

~/miniconda3/envs/pymc3/lib/python3.7/site-packages/jax/api.py in batched_fun(*args, **kwargs)
   1230         flat_fun, axis_name, axis_size, in_axes_flat,
   1231         lambda: flatten_axes("vmap out_axes", out_tree(), out_axes)
-> 1232     ).call_wrapped(*args_flat)
   1233     return tree_unflatten(out_tree(), out_flat)
   1234 

~/miniconda3/envs/pymc3/lib/python3.7/site-packages/jax/linear_util.py in call_wrapped(self, *args, **kwargs)
    164 
    165     try:
--> 166       ans = self.f(*args, **dict(self.params, **kwargs))
    167     except:
    168       # Some transformations yield from inside context managers, so we have to

~/miniconda3/envs/pymc3/lib/python3.7/site-packages/numpyro/infer/hmc.py in <lambda>(init_params, rng_key)
    500             model_args=model_args,
    501             model_kwargs=model_kwargs,
--> 502             rng_key=rng_key,
    503         )
    504         if rng_key.ndim == 1:

~/miniconda3/envs/pymc3/lib/python3.7/site-packages/numpyro/infer/hmc.py in init_kernel(init_params, num_warmup, step_size, inverse_mass_matrix, adapt_step_size, adapt_mass_matrix, dense_mass, target_accept_prob, trajectory_length, max_tree_depth, find_heuristic_step_size, forward_mode_differentiation, model_args, model_kwargs, rng_key)
    209         """
    210         step_size = lax.convert_element_type(step_size, jnp.result_type(float))
--> 211         trajectory_length = lax.convert_element_type(trajectory_length, jnp.result_type(float))
    212         nonlocal wa_update, max_treedepth, vv_update, wa_steps, forward_mode_ad
    213         forward_mode_ad = forward_mode_differentiation

~/miniconda3/envs/pymc3/lib/python3.7/site-packages/jax/_src/lax/lax.py in convert_element_type(operand, new_dtype, weak_type)
    455   else:
    456     return convert_element_type_p.bind(operand, new_dtype=new_dtype,
--> 457                                        weak_type=new_weak_type)
    458 
    459 def bitcast_convert_type(operand: Array, new_dtype: DType) -> Array:

~/miniconda3/envs/pymc3/lib/python3.7/site-packages/jax/core.py in bind(self, *args, **kwargs)
   1927     top_trace = find_top_trace(args)
   1928     if top_trace is None:
-> 1929       return self.impl(*args, **kwargs)
   1930 
   1931     tracers = map(top_trace.full_raise, args)

~/miniconda3/envs/pymc3/lib/python3.7/site-packages/jax/interpreters/xla.py in apply_primitive(prim, *args, **params)
    228 def apply_primitive(prim, *args, **params):
    229   """Impl rule that compiles and runs a single primitive 'prim' using XLA."""
--> 230   compiled_fun = xla_primitive_callable(prim, *unsafe_map(arg_spec, args), **params)
    231   return compiled_fun(*args)
    232 

~/miniconda3/envs/pymc3/lib/python3.7/site-packages/jax/interpreters/xla.py in arg_spec(x)
    220 
    221 def arg_spec(x):
--> 222   aval = abstractify(x)
    223   try:
    224     return aval, x._device

~/miniconda3/envs/pymc3/lib/python3.7/site-packages/jax/interpreters/xla.py in abstractify(x)
    169   if hasattr(x, '__jax_array__'):
    170     return abstractify(x.__jax_array__())
--> 171   raise TypeError(f"Argument '{x}' of type '{type(x)}' is not a valid JAX type")
    172 
    173 def _make_abstract_python_scalar(typ, _):

TypeError: Argument 'None' of type '<class 'NoneType'>' is not a valid JAX type
```

In any case, no immediate success with `fns` instead of `fns[0]`!

brandonwillard reviewed (changes requested):
See https://github.com/pymc-devs/pymc3/pull/4646.

twiecki commented:
Closing as it would need to be refactored.
